File: android/src/main/java/app/tauri/appauth/OAuthSessionPlugin.kt
// SPDX-License-Identifier: Apache-2.0

package app.tauri.appauth

import android.app.Activity
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.Plugin
import net.openid.appauth.AuthorizationService

/**
 * Tauri bridge over AppAuth-Android. Per-command bodies land in phase 5 of the
 * rewrite plan; for now every command rejects with `NOT_IMPLEMENTED` so the
 * native target compiles against the new dependency surface.
 */
@TauriPlugin
class OAuthSessionPlugin(activity: Activity) : Plugin(activity) {

    // Lazily instantiated by per-command handlers (phase 5) and disposed here
    // so Custom Tabs connections are released on activity teardown.
    private var authService: AuthorizationService? = null

    override fun onDestroy() {
        authService?.dispose()
        authService = null
        super.onDestroy()
    }

    @Command
    fun discover(invoke: Invoke) {
        invoke.reject("discover is not implemented yet", ERROR_NOT_IMPLEMENTED)
    }

    @Command
    fun register(invoke: Invoke) {
        invoke.reject("register is not implemented yet", ERROR_NOT_IMPLEMENTED)
    }

    @Command
    fun authorize(invoke: Invoke) {
        invoke.reject("authorize is not implemented yet", ERROR_NOT_IMPLEMENTED)
    }

    @Command
    fun authorizeBrowserOnly(invoke: Invoke) {
        invoke.reject("authorizeBrowserOnly is not implemented yet", ERROR_NOT_IMPLEMENTED)
    }

    @Command
    fun refresh(invoke: Invoke) {
        invoke.reject("refresh is not implemented yet", ERROR_NOT_IMPLEMENTED)
    }

    @Command
    fun endSession(invoke: Invoke) {
        invoke.reject("endSession is not implemented yet", ERROR_NOT_IMPLEMENTED)
    }

    private companion object {
        const val ERROR_NOT_IMPLEMENTED = "NOT_IMPLEMENTED"
    }
}

File: ios/Sources/OAuthSessionPlugin.swift
// SPDX-License-Identifier: Apache-2.0

import AppAuth
import Tauri
import UIKit

/// Tauri bridge over AppAuth-iOS. Per-command bodies land in phase 4 of the
/// rewrite plan; for now every command rejects with `NOT_IMPLEMENTED` so the
/// native target compiles against the new dependency surface.
class OAuthSessionPlugin: Plugin {
    /// Active AppAuth user-agent flow. Retained while a browser sheet or token
    /// exchange is outstanding so AppAuth can drive it to completion or
    /// cancellation.
    private var currentSession: OIDExternalUserAgentSession?

    @objc public func discover(_ invoke: Invoke) {
        invoke.reject("discover is not implemented yet", code: "NOT_IMPLEMENTED")
    }

    @objc public func register(_ invoke: Invoke) {
        invoke.reject("register is not implemented yet", code: "NOT_IMPLEMENTED")
    }

    @objc public func authorize(_ invoke: Invoke) {
        invoke.reject("authorize is not implemented yet", code: "NOT_IMPLEMENTED")
    }

    @objc public func authorizeBrowserOnly(_ invoke: Invoke) {
        invoke.reject("authorizeBrowserOnly is not implemented yet", code: "NOT_IMPLEMENTED")
    }

    @objc public func refresh(_ invoke: Invoke) {
        invoke.reject("refresh is not implemented yet", code: "NOT_IMPLEMENTED")
    }

    @objc public func endSession(_ invoke: Invoke) {
        invoke.reject("endSession is not implemented yet", code: "NOT_IMPLEMENTED")
    }
}

@_cdecl("init_plugin_oauth_session")
func initPlugin() -> Plugin {
    return OAuthSessionPlugin()
}
